pmd-java/src/test/java/net/sourceforge/pmd/typeresolution/ClassTypeResolverTest.java

```diff
@@ -775,22 +763,93 @@ public void testArrayTypes() throws JaxenException {
         // Make sure we got them all
         assertEquals("All expressions not tested", index, expressions.size());
     }
-    
+
+
+    @Test
+    public void testReferenceType() {
+        List<ASTReferenceType> expressions = selectNodes(ArrayTypes.class, ASTReferenceType.class);
+
+        int index = 0;
+
+        // int[] a = new int[1];
+        testSubtreeNodeTypes(expressions.get(index++), int[].class);
+
+        // Object[][] b = new Object[1][0];
+        testSubtreeNodeTypes(expressions.get(index++), Object[][].class);
+
+        // ArrayTypes[][][] c = new ArrayTypes[][][] { new ArrayTypes[1][2] };
+        testSubtreeNodeTypes(expressions.get(index++), ArrayTypes[][][].class);
+
+        // Make sure we got them all
+        assertEquals("All expressions not tested", index, expressions.size());
+    }
+
+
+    @Test
+    public void testHeterogeneousArrayFieldDeclaration() throws JaxenException {
+        List<ASTFieldDeclaration> fields = selectNodes(ArrayVariableDeclaration.class, ASTFieldDeclaration.class);
+        List<ASTLocalVariableDeclaration> locals = selectNodes(ArrayVariableDeclaration.class, ASTLocalVariableDeclaration.class);
+
+        // public int[] a, b[];
+        testPrimitiveTypeFieldDecl(fields.get(0));
+        testPrimitiveTypeFieldDecl(locals.get(0));
+
+        // public String[] c, d[];
+        testRefTypeFieldDecl(fields.get(1));
+        testRefTypeFieldDecl(locals.get(1));
+    }
+
+
+    // subtest
+    private void testPrimitiveTypeFieldDecl(Node declaration) throws JaxenException {
+
+        // public String[] c, d[];
```

Comment:
this comment is confusing, as this check is actually dealing with `public int[] a, b[];`.

Non the less, I find the whole method rather confusing. The name seems more generic than the implementation is (which actually matches down into the variable names themselves), and I'm not sure the xpath is the most readable / explicit way to test it.

If the xpath fails, the  `assertTrue` at line 812 will simply fail, but no clue is given to the developer as to which of the conditions the xpath tests for is failing (which is not minor, considering it's looking at the specific properties of 3 different nodes).

I know it's more succinct, but it's probably not the most future-proof way to check this.

Reply:
I fixed that, while avoiding some null checks for conciseness. It's not nearly as bad as I thought it would be in terms of code size, so thanks for pushing me to do that! XPath was the lazy solution